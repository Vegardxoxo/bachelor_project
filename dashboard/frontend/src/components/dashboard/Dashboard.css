.body {
    width: 100%;
}

#header {
    margin-top: 40px;
    display: flex;
    justify-content: center;
}

#boxes {
    margin-top: 20px;
    text-align: center;
    display: flex;
    justify-content: center;
}

#card {
    width: 300px;
    height: 180px;
}

#headline {
    margin-top: 15px;
    margin-bottom: 15px;
}

#cardTitle {
    color: #002d53;
    padding-top: 10px;
    padding-left: 5px;
    text-align: center;
    font-family: Source Sans Pro, sans-serif;
    font-size: 20pt;
    font-weight: 400;
}

#numbersBoxes {
    color: #002d53;
    padding-top: 10px;
    text-align: center;
    font-family: Source Sans Pro, sans-serif;
    font-size: 35px;
    font-weight: lighter;
}

#organization {
    text-align: left;
    font-family: Source Sans Pro, sans-serif;
    margin-top: 20px;
    display: flex;
    justify-content: flex-start;
    font-weight: 400;
    font-size: 30px;
}

#org_name {
    font-family: Source Sans Pro, sans-serif;
    font-size: 30pt;
    font-weight: 400;
    display: flex;
    width: 780px;
}

#selector {
    padding-top: 40px;
    position: relative;
    display: flex;
    justify-content: flex-end;
}

#donut_chart {
    display: flex;
    justify-content: center;

}

#portal-link {
    color: #005aa7;
    text-decoration: none;
    font-family: Source Sans Pro, sans-serif;
    font-style: normal;
    font-weight: 400;
    font-size: 30px;
    line-height: 43px;
    display: flex;
    align-items: center;
    gap: 8px;
    margin-bottom: 10px;
}

.license-portal {
    justify-content: flex-end;
    align-content: flex-end;
}

#portal-link:hover {
    border-bottom: 2px solid #005aa7;
}