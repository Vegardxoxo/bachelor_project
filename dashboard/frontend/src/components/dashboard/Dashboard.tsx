import React, {useEffect, useState} from 'react';
import './Dashboard.css';
import {Grid, Stack, Typography} from '@mui/material';
import InfoBox from './InfoBox/InfoBox';
import DonutChart from './DonutChart/DonutChart';
import ActiveLastBreadcrumb from './ActivateLastBreadcrumb/ActivateLastBreadcrumb';
import {LeaderboardBox} from './LeaderboardBox/LeaderboardBox';
import {useRecoilValue} from 'recoil';
import {orgAtom} from '../../globalVariables/variables';
import {fetchInfoBoxData} from '../../api/calls';
import MuiLoadingSpinner from '../spinner/MuiLoadingSpinner';
import LogoutOutlinedIcon from '@mui/icons-material/LogoutOutlined';
import {SavingsBox} from './SavingsBox/SavingsBox';
import {Link} from 'react-router-dom';
import {Count} from '../../Interfaces';

/**
 * Renders the dashboard page.
 */
function Dashboard() {

    // Retrieve organization name from global state
    const org = useRecoilValue(orgAtom);

    // Define state variables for box data and access token
    const [boxData, setBoxData] = useState<Count[] | undefined>(undefined);
    const accessToken = localStorage.getItem('access');

    // Fetch box data
    useEffect(() => {
        const fetchData = async () => {
            try {
                if (accessToken && org) {
                    const boxDataResponse: Count[] | undefined = await fetchInfoBoxData(org);
                    if (boxDataResponse !== undefined) {
                        setBoxData(boxDataResponse);
                    }
                }
            } catch (error) {
                console.error('Error while fetching data:', error);
            }
        };

        fetchData();
    }, [accessToken, org]);


    return (
        <>
            {boxData ? ( // If box data has been fetched
                <div className={'body'}>
                    <Grid container id="header">
                        <Stack spacing={2}>
                            <Grid item sx={{marginLeft: -21, marginTop: -5}}>
                                {/* Adds breadcrumbs to the top left of the page */}
                                <ActiveLastBreadcrumb/>
                            </Grid>
                            <Grid container>
                                <Stack direction="row">
                                    <Typography id="org_name">{org}</Typography>
                                    <Stack direction="row">
                                        <Link to="/lisensportal" id="portal-link">
                                            Til lisensportalen
                                            <LogoutOutlinedIcon style={{alignContent: 'center'}}/>
                                        </Link>
                                    </Stack>
                                </Stack>
                            </Grid>
                        </Stack>
                    </Grid>
                    <Grid container id="boxes">
                        <Grid item>
                            <Stack direction="row" spacing={8}>
                                {/* Display info boxes with data */}
                                <InfoBox
                                    title="Totale Lisenser"
                                    data-testid="totale-lisenser-id"
                                    numberOfLicenses={boxData[0]?.total_licenses ?? 0}
                                />
                                <InfoBox
                                    title="Ubrukte Lisenser"
                                    numberOfLicenses={boxData[0]?.never_used ?? 0}
                                />
                                <InfoBox
                                    title="Ledige Lisenser"
                                    numberOfLicenses={boxData[0]?.unused_licenses ?? 0}
                                />
                            </Stack>
                        </Grid>
                    </Grid>
                    <Grid container id="boxes">
                        <Grid container id={'donut_chart'}>
                            {/* Display donut chart with data */}
                            <DonutChart
                                never_used={boxData[0].never_used}
                                total_licenses={boxData[0].total_licenses}
                                unused_licenses={boxData[0].unused_licenses}
                                active_licenses={boxData[0].active_licenses}
                                width={670}
                                height={425}
                                showInformation={true}
                                title='Total oversikt'
                            />
                            {/* Stack component containing SavingsBox and LeaderboardBox */}
                            <Grid item sx={{ml: 8, mt: 7}}>
                                <Stack direction={'column'} spacing={4.38}>
                                    <SavingsBox/>
                                    <LeaderboardBox/>
                                </Stack>
                            </Grid>
                        </Grid>
                    </Grid>
                </div>
            ) : (
                /* If the page is not loaded, a spinner is displayed. */
                <MuiLoadingSpinner/>
            )}
        </>
    );
}

export default Dashboard;
