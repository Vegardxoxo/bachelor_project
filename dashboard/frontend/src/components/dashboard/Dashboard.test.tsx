import React from 'react';
import {render} from '@testing-library/react';
import {MemoryRouter} from 'react-router-dom';
import {RecoilRoot} from 'recoil';
import Dashboard from "./Dashboard";
import renderer from "react-test-renderer";

describe('Dashboard', () => {
    it('renders without crashing', () => {
        render(
            <RecoilRoot>
                <MemoryRouter>
                    <Dashboard/>
                </MemoryRouter>
            </RecoilRoot>
        );
    });

    it('matches snapshot', async () => {
        const testRenderer = renderer.create(
            <RecoilRoot>
                <MemoryRouter>
                    <Dashboard/>
                </MemoryRouter>
            </RecoilRoot>
        );
        expect(testRenderer.toJSON()).toMatchSnapshot();
    });


})