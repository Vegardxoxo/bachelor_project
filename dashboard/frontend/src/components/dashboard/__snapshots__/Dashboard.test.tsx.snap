// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard matches snapshot 1`] = `
<div
  className="MuiBox-root css-1lb6n3v"
>
  <span
    className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
    role="progressbar"
    style={
      Object {
        "height": 80,
        "width": 80,
      }
    }
  >
    <svg
      className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
        cx={44}
        cy={44}
        fill="none"
        r={20.2}
        strokeWidth={3.6}
        style={Object {}}
      />
    </svg>
  </span>
</div>
`;
